src/Visitor/MutationsCollectorVisitor.php

```diff
@@ -63,6 +63,10 @@ public function __construct(
     public function leaveNode(Node $node)
     {
         foreach ($this->mutators as $mutator) {
+            if (!$mutator->shouldMutate($node)) {
```

Comment:
I would move it below the next 2 `if`s because `shouldMutate()` sometimes is quite expensive operation, for example see this one that uses reflection and iterates over parent classes/interfaces: https://github.com/infection/infection/blob/master/src/Mutator/FunctionSignature/PublicVisibility.php#L68

On the other side, things like  `isOnFunctionSignature` and `$node->getAttribute(ReflectionVisitor::IS_INSIDE_FUNCTION_KEY)` are just getters for already calculated values

Reply:
I've run this change under `perf`, and I've run it again against the Infection codebase, and on five runs `perf` showed me average run time of 26.63 second for the currently implement variant, and 33.44 seconds for the variant you're suggestion. In both cases the standard deviation was about 5% or about 2 seconds.

Do you still think we have to move it below?

Reply:
We can micro-optimize here if we distinguish fast and slow mutators by their parent class or else. I'm not sure it's the right time, though.

Reply:
If we are looking at those two suspects only:

```bash
vendor/bin/phpunit --coverage-xml=coverage/coverage-xml --log-junit=coverage/phpunit.junit.xml
perf stat -r 20 php bin/infection --quiet --coverage=coverage --threads=12 --mutators=ProtectedVisibility,PublicVisibility
```
For a case where `shouldMutate` comes first I get: 2.234926127 seconds ( +-  0,21% )
For a case where `shouldMutate` goes below next 2 `if`s: 2.292622426 seconds  ( +-  0,29% )

SD is about same in both cases, and difference is about 2%. Which makes me think FunctionSignature mutators aren't the bottleneck here...

Reply:
Thanks for the explanation. Probably the bottleneck is `$mutator->isIgnored()` function. Have you tried to move `shouldMutate()` after the next `if` but before the `if` with `isIgnored()`?

Anyway, I'm happy to merge as is.

Reply:
My poor man's instrumentation shows that there isn't a single true bottleneck. 

| Method                | ms per 1000  | 
| --------------------- | -------------|
| isCoveredByTest       | 1.08         | 
| isIgnored             | 0.99         | 
| isOnFunctionSignature | 0.49         |
| shouldMutate          | 0.348        |

What `leaveNode` does is a negative selection.

Therefore, only having average timings we can make a decision: the fastest function to exclude a given mutator should come first, and slowest the last. This way the negative selection will happen in the most efficient way.

It appears that by sheer luck or providence the current layout is optimal.

Reply:
Great, thank you!

Reply:
What I also discovered is that `MutationsCollectorVisitor` gets constructed a lot of times. Probably once per each thread. Therefore we can get at least partial speed-up if we reuse result of MutationsCollectorVisitor in further invocations. It seems to remain a constant between invocations.